oauth2/src/main/java/com/linecorp/armeria/internal/common/auth/oauth2/AbstractOAuth2Request.java

```diff
@@ -0,0 +1,224 @@
+                                    @Nullable ClientAuthorization clientAuthorization) {
+        this.endpoint = Objects.requireNonNull(endpoint, "endpoint");
+        this.endpointPath = Objects.requireNonNull(endpointPath, "endpointPath");
+        this.clientAuthorization = clientAuthorization; // optional
+    }
+
+    /**
+     * Returns the {@link WebClient} of the authorization endpoint.
+     */
+    protected final WebClient endpoint() {
+        return endpoint;
+    }
+
+    /**
+     * Returns the authorization endpoint path.
+     */
+    protected final String endpointPath() {
+        return endpointPath;
+    }
+
+    /**
+     * Returns the client authorization object.
+     */
+    @Nullable
+    protected final ClientAuthorization clientAuthorization() {
+        return clientAuthorization;
+    }
+
+    /**
+     * Extracts data from OK response and converts it to the target type {@code T}.
+     */
+    protected abstract T extractOkResults(AggregatedHttpResponse response, QueryParams requestFormData);
+
+    /**
+     * Returns the value for the {@link HttpHeaderNames#AUTHORIZATION}.
+     */
+    @Nullable
+    protected final String authorizationHeaderValue() {
+        return clientAuthorization == null ? null : clientAuthorization.asHeaderValue();
+    }
+
+    /**
+     * Sets client credentials as form data parameters,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     */
+    protected final void addCredentialsAsBodyParameters(QueryParamsBuilder formBuilder) {
+        if (clientAuthorization != null) {
+            clientAuthorization.addAsBodyParameters(formBuilder);
+        }
+    }
+
+    /**
+     * Makes a request to the authorization endpoint using supplied {@code requestForm} parameters and converts
+     * the result to the given type {@code T}.
+     */
+    protected CompletableFuture<T> executeWithParameters(QueryParams requestFormData) {
+        final HttpResponse response = endpoint().execute(createHttpRequest(endpointPath, requestFormData));
+        // when response aggregated, then extract the results...
+        return response.aggregate().thenApply(r -> extractResults(r, requestFormData));
+    }
+
+    /**
+     * Produces {@link HttpRequest} based on this object.
+     */
+    private HttpRequest createHttpRequest(String endpointPath, QueryParams requestFormData) {
+        final RequestHeadersBuilder headersBuilder =
+                RequestHeaders.of(HttpMethod.POST, endpointPath).toBuilder();
+        final String authorizationHeaderValue = authorizationHeaderValue();
+        if (authorizationHeaderValue != null) {
+            headersBuilder.add(HttpHeaderNames.AUTHORIZATION, authorizationHeaderValue);
+        } else {
+            requestFormData = requestFormData.withMutations(this::addCredentialsAsBodyParameters);
+        }
+        headersBuilder.add(HttpHeaderNames.CONTENT_TYPE, MediaType.FORM_DATA.toString());
```

Comment:
A `MediaType` will be converted to a `String` in `StringValueConverter`.
```suggestion
        headersBuilder.addObject(HttpHeaderNames.CONTENT_TYPE, MediaType.FORM_DATA);
```

Reply:
Thought this could handle only a fine set of the specific classes like `Date` or `Instant`.
Modified it as per your suggesting, assuming it works with generic objects.

Reply:
`MediaType` is one of the supported types, so no worries :-)